content/lib/Pdf/Content/UnicodeCMap.hs

```diff
@@ -66,11 +66,13 @@ unicodeCMapNextGlyph cmap = go 1
       else if any (inRange glyph) (unicodeCMapCodeRanges cmap)
              then Just (toCode glyph, ByteString.drop n str)
              else go (n + 1) str
-  inRange glyph (start, end) = glyph >= start && glyph <= end
+  inRange glyph (start, end)
+    = ByteString.length glyph == ByteString.length start
+    && glyph >= start && glyph <= end
```

Comment:
This makes the function eat the right number of bytes. Yes, I thought of that, too. 

But I think eating the right number of bytes is not enough. Also the right decoding function has be used for the eaten bytes. Characters in the CMap are always Utf16BE, but not characters in the text. This function eats the bytes from the text.

The text you extracted from the GoogleBook using this fix is correct, yes. But that's because the decoding using `fromEnum . toCode` works for UCS-2 and Utf16BE, and we have UCS-2 in this PDF. 

But `toCode` is only a 1:1-mapping of the bytes from the bytestring to codepoints. If you take a PDF with text encoded with UTF8 and characters represented with multiple bytes, the decoding problem would become visible.

Reply:
Hmm. I think we have quite different understanding how unicode cmap works. I'm not sure which one is correct, so let me explain my understanding.

Text in PDF file doesn't have any encoding belong what is defined in the cmap. Basically it's just a sequence of bytes. Using `codespacerange` we can split the sequence into individual "characters". Though they are not actually characters, just unique identifiers of glyphs in the font. Then we use `bfchar` and `bfrange` to map these glyph identifiers to unicode code points. We never apply any encoding to the glyph identifier. In fact it makes no sense to say that characters in the text itself have any encoding, neither ucs-2, nor utf8.

In the existing code glyph identifiers are represented as `Int`. The `unicodeCMapNextGlyph` returns the next glyph identifier and the rest of the text. And `unicodeCMapDecodeGlyph` maps the identifier to the unicode code point. It's done in either using `bfchar` or using `bfragen`. The former case is trivial - we just use whatever unicode codepoint is specified to the particular glyph identifier in the `bfchar` section. The later case is a bit more complicated, we find the offset of the identifier in the range and then use the specified unicode codepoint + the offset, that's exactly what fromEnum\toEnum part does. Se we never assume any particular encoding in the input text, and I don't see why we should do that.

It's totally possible that I'm completely wrong here. I'd appreciate if you'll correct me.

Regarding the "if you take a PDF with text encoded with UTF8 and characters represented with multiple bytes" part. I understand what it means to represent characters with multiple bytes. But as I described above the notion of text encoding here doesn't make sense for me.

Reply:
Ok, after looking in den specs again, I got uncertain about my argument. Since your understanding is quite clear and your fix is neat and works for me, I'm in favour of your fix.